src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -1,235 +1,277 @@
+                             :sql                       (sql.qp/format-honeysql driver update-hsql)
+                             :status-code               400})))
+          {:rows-updated [1]})
+        (catch Exception e
+          (let [e-data (if (::incorrect-number-updated (ex-data e))
+                         (ex-data e)
+                         (parse-error driver conn e))]
+            (throw
+             (ex-info (or (ex-message e) "Update action error.")
+                      (assoc e-data :status-code 400)))))))))
+
+(defn- select-created-row
+  "H2 and MySQL are dumb and `RETURN_GENERATED_KEYS` only returns the ID of the newly created row. This function will
+  `SELECT` the newly created row."
+  [driver raw-hsql conn result]
+  (let [select-hsql     (assoc raw-hsql
+                               :select [:*]
+                               ;; :and with a single clause will be optimized in HoneySQL
+                               :where (into [:and]
+                                            (for [[col val] result]
+                                              [:= (keyword col) val])))
+        select-sql-args (sql.qp/format-honeysql driver select-hsql)]
+    (log/tracef ":row/create SELECT HoneySQL:\n\n%s" (u/pprint-to-str select-hsql))
+    (log/tracef ":row/create SELECT SQL + args:\n\n%s" (u/pprint-to-str select-sql-args))
+    (first (jdbc/query {:connection conn} select-sql-args {:identifiers identity, :transaction? false}))))
 
 (defmethod actions/perform-action!* [:sql-jdbc :row/create]
   [driver _action _database {database-id :database :keys [create-row] :as query}]
-  (let [conn        (sql-jdbc.conn/db->pooled-connection-spec database-id)
-        raw-hsql    (qp.store/with-store
+  (let [raw-hsql    (qp.store/with-store
                       (try
                         (qp/preprocess query) ; seeds qp store as a side effect so we can generate honeysql
                         (sql.qp/mbql->honeysql driver query)
                         (catch Exception e
                           (catch-throw e 404))))
         create-hsql (-> raw-hsql
                         (assoc :insert-into (first (:from raw-hsql)))
-                        (assoc :values [(cast-values create-row (get-in query [:query :source-table]))])
-                        (dissoc :select :from))]
-    {:created-row
-     (if (= driver :postgres)
-       ;; postgres is happy with "returning *", but mysql and h2 aren't so just use select *.
-       (let [pg-create-hsql (-> create-hsql (assoc :returning [:*]))
-             raw-sql        (hformat/format pg-create-hsql)]
-         (try (jdbc/execute! conn raw-sql {:return-keys true})
-              (catch Exception e
-                (throw
-                 (ex-info "Create action error." (assoc (parse-error driver conn e) :status-code 400))))))
-       (try (let [col->val (jdbc/execute! conn (hformat/format create-hsql) {:return-keys true})]
```

Comment:
You can see both

- #23932
- #23927

on this line; those are both fixed in the new version